packages/snaps-rpc-methods/src/permitted/openWebSocket.test.ts

```diff
@@ -0,0 +1,144 @@
+import { JsonRpcEngine } from '@metamask/json-rpc-engine';
+import type { OpenWebSocketResult } from '@metamask/snaps-sdk';
+import type { JsonRpcRequest, PendingJsonRpcResponse } from '@metamask/utils';
+
+import type { OpenWebSocketParameters } from './openWebSocket';
+import { openWebSocketHandler } from './openWebSocket';
+
+describe('snap_openWebSocket', () => {
+  describe('openWebSocketHandler', () => {
+    it('has the expected shape', () => {
+      expect(openWebSocketHandler).toMatchObject({
+        methodNames: ['snap_openWebSocket'],
+        implementation: expect.any(Function),
+        hookNames: {
+          hasPermission: true,
+          openWebSocket: true,
+        },
+      });
+    });
+  });
+
+  describe('implementation', () => {
+    it('throws if the origin does not have permission', async () => {
+      const { implementation } = openWebSocketHandler;
+
+      const openWebSocket = jest.fn();
+      const hasPermission = jest.fn().mockReturnValue(false);
+      const hooks = { hasPermission, openWebSocket };
+
+      const engine = new JsonRpcEngine();
+
+      engine.push((request, response, next, end) => {
+        const result = implementation(
+          request as JsonRpcRequest<OpenWebSocketParameters>,
+          response as PendingJsonRpcResponse<OpenWebSocketResult>,
+          next,
+          end,
+          hooks,
+        );
+
+        result?.catch(end);
+      });
+
+      const response = await engine.handle({
+        jsonrpc: '2.0',
+        id: 1,
+        method: 'snap_openWebSocket',
+        params: {
+          url: 'wss://metamask.io',
+        },
+      });
+
+      expect(response).toStrictEqual({
+        error: {
+          code: 4100,
+          message:
+            'The requested account and/or method has not been authorized by the user.',
+          stack: expect.any(String),
+        },
+        id: 1,
+        jsonrpc: '2.0',
+      });
+    });
+
+    it('throws if invalid parameters are passed', async () => {
+      const { implementation } = openWebSocketHandler;
+
+      const openWebSocket = jest.fn();
+      const hasPermission = jest.fn().mockReturnValue(true);
+      const hooks = { hasPermission, openWebSocket };
+
+      const engine = new JsonRpcEngine();
+
+      engine.push((request, response, next, end) => {
+        const result = implementation(
+          request as JsonRpcRequest<OpenWebSocketParameters>,
+          response as PendingJsonRpcResponse<OpenWebSocketResult>,
+          next,
+          end,
+          hooks,
+        );
+
+        result?.catch(end);
+      });
+
+      const response = await engine.handle({
+        jsonrpc: '2.0',
+        id: 1,
+        method: 'snap_openWebSocket',
+        params: {
+          url: 'ws://metamask.io',
+        },
+      });
+
+      expect(response).toStrictEqual({
+        error: {
+          code: -32602,
+          message:
+            'Invalid params: At path: url -- Expected URL, got "ws://metamask.io"..',
```

Comment:
The double periods are weird 😅 I know we do this in some other places, but IIRC we have a util function somewhere to get an error message, removing the double period.
```suggestion
            'Invalid params: At path: url -- Expected URL, got "ws://metamask.io".',
```

Reply:
The error message is also a bit confusing, since `ws://metamask.io` is a valid URL. I think it should mention that we only support `wss://`.

Reply:
Yeah, the struct we have for this isn't great. Unsure if we want to blow up the scope more for this PR to improve 🤔